File: libs/as-sdk/assembly/process.ts
import {
	lazyArgv,
	lazyEnv,
	wasi_process,
} from "@assemblyscript/wasi-shim/assembly/wasi_process";
import { execution_result } from "./bindings/seda_v1";
import { Bytes } from "./bytes";
import { VM_MODE_DR, VM_MODE_ENV_KEY, VM_MODE_TALLY } from "./vm-modes";

const POSIX_SUCCESS_CODE = u8(0);
const POSIX_ERROR_CODE = u8(1);

const argv = lazyArgv();
const env = lazyEnv();

/**
 * A set of methods to interact with the process in which the Oracle Program is running.
 *
 * @category Program
 */
export default class Process {
	/** @hidden */
	private constructor() {}

	/**
	 * Gets all the input arguments from the Data Request
	 * First argument (index: 0) is the Oracle Program ID
	 * Second argument (index: 1) is the input of the Data Request
	 *
	 * @returns {string[]} An array of input arguments
	 * @example
	 * ```ts
	 * const args = Process.args();
	 * ```
	 */
	static args(): string[] {
		return argv;
	}

	/**
	 * Gets all the environment variables as a Map
	 *
	 * @returns {Map<string, string>} key, value pair with all environment variables
	 * @example
	 * ```ts
	 * const env = Process.env();
	 *
	 * const vmMode = env.get('VM_MODE');
	 * ```
	 */
	static envs(): Map<string, string> {
		return env;
	}

	/**
	 * Gets the data request / tally inputs
	 *
	 * @returns {Bytes} bytes encoded inputs
	 * @example
	 * ```ts
	 * const inputs = Process.getInputs();
	 * const inputAsString = inputs.toUtf8String();
	 *
	 * console.log(inputAsString);
	 * ```
	 */
	static getInputs(): Bytes {
		// Data at index 1 is the dr/tally inputs encoded as hex
		const data = Process.args().at(1);

		return Bytes.fromHexString(data);
	}

	/**
	 * Gets the mode of the VM instance.
	 *
	 * @returns {string} The mode of the VM, either 'dr' or 'tally'.
	 */
	static getVmMode(): string {
		const env = Process.envs();

		return env.get(VM_MODE_ENV_KEY);
	}

	/**
	 * Returns true when the VM instance is in 'tally' mode.
	 *
	 * @returns {boolean}
	 */
	static isTallyVmMode(): boolean {
		return Process.getVmMode() === VM_MODE_TALLY;
	}

	/**
	 * Returns true when the VM instance is in 'dr' mode.
	 *
	 * @returns {boolean}
	 */
	static isDrVmMode(): boolean {
		return Process.getVmMode() === VM_MODE_DR;
	}

	/**
	 * Exits the process (no result set)
	 *
	 * @param {u8} code Exit code (POSIX compatible, 0 is success, >= 1 is error)
	 * @example
	 * ```
	 * Process.exit(0);
	 * ```
	 */
	static exit(code: u8): void {
		exitWithResult(code, Bytes.empty());
	}

	/**
	 * Exits the process with a bytes encoded result.
	 * This sets Data Request execution result to the bytes
	 *
	 * @param {Bytes} result Bytes encoded result, which will be sent back to the contract
	 * @example
	 * ```ts
	 * const result = "{\"price\": \"10.23\"}";
	 *
	 * Process.success(Bytes.fromUtf8String(result));
	 * ```
	 */
	static success(result: Bytes): void {
		exitWithResult(POSIX_SUCCESS_CODE, result);
	}

	/**
	 * Exits the process with a bytes encoded result.
	 * This sets Data Request execution result to the bytes
	 *
	 * @param {Bytes} result Bytes encoded result, which will be sent back to the contract
	 * @param {u8} [code=1] code Exit code, defaults to 1 (POSIX compatible, 0 is success, >= 1 is error)
	 * @example
	 * ```ts
	 * const error = "Failed to fetch data from https://example.com";
	 *
	 * Process.error(Bytes.fromUtf8String(error));
	 * ```
	 */
	static error(result: Bytes, code: u8 = POSIX_ERROR_CODE): void {
		exitWithResult(code, result);
	}
}

function exitWithResult(code: u8, result: Bytes): void {
	const buffer = result.value.buffer;
	const bufferPtr = changetype<usize>(buffer);

	execution_result(bufferPtr, buffer.byteLength);
	wasi_process.exit(u32(code));
}

File: libs/rs-sdk/sdk/src/process.rs
//! Process management utilities for the SDK.
//!
//! This module provides functionality for handling program arguments, environment variables,
//! and process termination with specific exit codes and results.
//!
//! # Examples
//!
//! ```no_run
//! use seda_sdk::process::Process;
//!
//! // Handle success case
//! let result = vec![1, 2, 3];
//! Process::success(&result);
//!
//! // Handle error case with message
//! Process::exit_with_message(1, "Operation failed");
//! ```

use std::{
    collections::BTreeMap,
    env::{self},
    process,
};

use crate::{
    raw::{self},
    vm_modes::{DR_REPLICATION_FACTOR_ENV_KEY, VM_MODE_DR, VM_MODE_ENV_KEY, VM_MODE_TALLY},
};

/// Utilities for managing process-level operations.
pub struct Process;

impl Process {
    /// Returns a vector of command-line arguments passed to the program.
    ///
    /// # Examples
    ///
    /// ```no_run
    /// use seda_sdk::process::Process;
    ///
    /// let args = Process::args();
    /// // args[0] is the program name
    /// // args[1] would be the first argument
    /// ```
    pub fn args() -> Vec<String> {
        env::args().collect()
    }

    /// Returns a map of all environment variables.
    ///
    /// # Examples
    ///
    /// ```no_run
    /// use seda_sdk::process::Process;
    ///
    /// let env_vars = Process::envs();
    /// for (key, value) in env_vars {
    ///     println!("{}: {}", key, value);
    /// }
    /// ```
    pub fn envs() -> BTreeMap<String, String> {
        env::vars().collect()
    }

    /// Retrieves and decodes the data request inputs
    ///
    /// # Examples
    ///
    /// ```no_run
    /// use seda_sdk::process::Process;
    ///
    /// let input_bytes = Process::get_inputs();
    /// // Process the input bytes...
    /// ```
    pub fn get_inputs() -> Vec<u8> {
        // 1 is inputs because 0 is name of the program
        let args = Self::args();
        let raw_input = args.get(1).expect("Arg at index 1 must best set");

        hex::decode(raw_input).expect("Arg at index 1 is not a valid hex string")
    }

    /// Gets the VM mode from environment variables.
    fn get_vm_mode() -> String {
        env::var(VM_MODE_ENV_KEY).expect("VM_MODE is not set in environment")
    }

    /// Checks if the current VM mode is set to tally mode.
    ///
    /// # Examples
    ///
    /// ```no_run
    /// use seda_sdk::process::Process;
    ///
    /// if Process::is_tally_vm_mode() {
    ///     // Handle tally mode specific logic
    /// }
    /// ```
    pub fn is_tally_vm_mode() -> bool {
        Self::get_vm_mode() == VM_MODE_TALLY
    }

    /// Checks if the current VM mode is set to DR mode.
    ///
    /// # Examples
    ///
    /// ```no_run
    /// use seda_sdk::process::Process;
    ///
    /// if Process::is_dr_vm_mode() {
    ///     let replication = Process::replication_factor();
    ///     // Handle DR mode specific logic
    /// }
    /// ```
    pub fn is_dr_vm_mode() -> bool {
        Self::get_vm_mode() == VM_MODE_DR
    }

    /// Exits the process successfully (code 0) with the given result.
    ///
    /// # Examples
    ///
    /// ```no_run
    /// use seda_sdk::process::Process;
    ///
    /// let result = vec![0x01, 0x02, 0x03];
    /// Process::success(&result);
    /// ```
    pub fn success(result: &[u8]) {
        Self::exit_with_result(0, result);
    }

    /// Exits the process with an error (code 1) and the given result.
    ///
    /// # Examples
    ///
    /// ```no_run
    /// use seda_sdk::process::Process;
    ///
    /// let error_data = vec![0xFF];
    /// Process::error(&error_data);
    /// ```
    pub fn error(result: &[u8]) {
        Self::exit_with_result(1, result);
    }

    /// Gets the replication factor for the data request
    ///
    /// # Examples
    ///
    /// ```no_run
    /// use seda_sdk::process::Process;
    ///
    /// if Process::is_dr_vm_mode() {
    ///     let factor = Process::replication_factor();
    ///     println!("Replication factor: {}", factor);
    /// }
    /// ```
    pub fn replication_factor() -> u16 {
        let variable = env::var(DR_REPLICATION_FACTOR_ENV_KEY).unwrap_or_else(|_| {
            panic!("{DR_REPLICATION_FACTOR_ENV_KEY} is not set in environment")
        });

        variable
            .parse()
            .unwrap_or_else(|_| panic!("{DR_REPLICATION_FACTOR_ENV_KEY} must be a valid u16"))
    }

    /// Exits the process with the given code and message.
    ///
    /// # Examples
    ///
    /// ```no_run
    /// use seda_sdk::process::Process;
    ///
    /// // Exit with an error message
    /// Process::exit_with_message(1, "Operation failed: invalid input");
    ///
    /// // Exit with success message
    /// Process::exit_with_message(0, "Operation completed successfully");
    /// ```
    pub fn exit_with_message(code: u8, message: &str) {
        let message_encoded = message.as_bytes();
        Self::exit_with_result(code, message_encoded);
    }

    /// Exits the process with the given code and result bytes.
    ///
    /// # Examples
    ///
    /// ```no_run
    /// use seda_sdk::process::Process;
    ///
    /// let result = vec![1, 2, 3, 4];
    /// Process::exit_with_result(0, &result);
    /// ```
    pub fn exit_with_result(code: u8, result: &[u8]) {
        unsafe { raw::execution_result(result.as_ptr(), result.len() as u32) };
        process::exit(code.into());
    }

    /// Exits the process with the given code and an empty result.
    ///
    /// # Examples
    ///
    /// ```no_run
    /// use seda_sdk::process::Process;
    ///
    /// // Exit successfully with no result
    /// Process::exit(0);
    ///
    /// // Exit with error code
    /// Process::exit(1);
    /// ```
    pub fn exit(code: u8) {
        Self::exit_with_result(code, &[]);
    }
}
